prometheus/src/test/java/com/uber/m3/tally/prometheus/PrometheusReporterTest.java

```diff
@@ -0,0 +1,906 @@
+            reporter.reportGauge("test", tags, 23);
+            Double metricValue = getMetricSample(registry, "test", tags, null);
+            Assert.assertThat(metricValue, is(23d));
+
+            reporter.reportGauge("test", tags, 42);
+            metricValue = getMetricSample(registry, "test", tags, null);
+            Assert.assertThat(metricValue, is(42d));
+            // make sure that gauge with the same tag keys registered only ones
+            Mockito.verify(registry, times(1)).register(Matchers.any());
+
+            // make sure that gauge with the same tag keys but different tags values registered only ones but report
+            // separate metrics.
+            tags.put("a", "2");
+            reporter.reportGauge("test", tags, 13);
+            metricValue = getMetricSample(registry, "test", tags, null);
+            Assert.assertThat(metricValue, is(13d));
+        }
+
+        @Test(expected = IllegalArgumentException.class)
+        public void registeringGaugeWithSameNameButDifferentTagsShouldFail() {
+            Map<String, String> tags = new HashMap<>();
+            tags.put("a", "1");
+            tags.put("b", "1");
+            reporter.reportGauge("test", tags, 23);
+            tags.put("c", "1");
+            reporter.reportGauge("test", tags, 19);
+        }
+
+        @Test
+        public void gaugesWithDifferentNamesAndSameTagsRegisteredSeparately() {
+            Map<String, String> tags = new HashMap<>();
+            tags.put("a", "1");
+            tags.put("b", "1");
+            reporter.reportGauge("test1", tags, 23);
+            reporter.reportGauge("test2", tags, 42);
+            Double metricValue1 = getMetricSample(registry, "test1", tags, null);
+            Assert.assertThat(metricValue1, is(23d));
+
+            Double metricValue2 = getMetricSample(registry, "test2", tags, null);
+            Assert.assertThat(metricValue2, is(42d));
+            Mockito.verify(registry, times(2)).register(Mockito.any());
+        }
+    }
+
+    @RunWith(JUnit4.class)
+    public static class ReportTimerTest {
+        private CollectorRegistry registry;
+        private PrometheusReporter reporterSummary;
+
+        @Before
+        public void init() {
+            registry = Mockito.spy(new CollectorRegistry(true));
+            reporterSummary = PrometheusReporter.builder()
+                    .registry(registry)
+                    .timerType(TimerType.SUMMARY)
+                    .build();
+        }
+
+        @Test
+        public void reportTimerNoTags() {
+            reporterSummary.reportTimer("test", null, Duration.ofSeconds(42));
+            Double metricValue = getMetricSample(registry, "test_count", null, null);
+            Assert.assertThat(metricValue, is(1d));
+            Mockito.verify(registry, times(1)).register(Mockito.any());
+        }
+
+        @Test
+        public void reportTimerWithTags() {
+            Map<String, String> tags = singletonMap("key", "value");
```

Comment:
Seems like these tests are fairly similar to each other--think we could table test them or factor out helpers? Basic form seems to be something like:

- Report a metric
- Assert the metric was registered and has the correct value

Reply:
yeah, i tried to group them as table tests, but ended up with weird unreadable functional abstractions, cause `report<Something> ` methods accept different types and number of arguments. I tried to factor out common util functionalities behind `getMetricSample` and `assertBucket` methods and group all the tests in separate test-classes, which i find the most readable among my other attempts to unify these tests